### PR TITLE
Improve tile lifecycle determinism 

### PR DESCRIPTION
**Pull request by mwilsnd**
Addresses some observed behavior in the lifetime of tiles that wasn't initially expected. This PR makes some minor changes to how GeometryTiles update their pending status, to better reflect what the tile is actually doing. One of the more interesting observations is how a tile pyramid's call to `setLayers()` can dispatch a task on the worker thread before the tile has even loaded.

This PR should address #2807

### CODE REVIEW

**Comment by louwers** [comment]
Looks like it will also fix #2826!

File: src/mbgl/tile/geometry_tile.cpp [hunk 1]
```diff
@@ -245,11 +245,13 @@ void GeometryTile::setData(std::unique_ptr<const GeometryTileData> data_) {
 void GeometryTile::reset() {
     MLN_TRACE_FUNC();
 
+    if (pending) {
+        observer->onTileAction(id, sourceID, TileOperation::Cancelled);
+        pending = false;
+    }
+
     // Mark the tile as pending again if it was complete before to prevent
     // signaling a complete state despite pending parse operations.
```
Comment (louwers) [inline src/mbgl/tile/geometry_tile.cpp:254]: Is this comment still accurate?

File: test/map/map.test.cpp [hunk 2]
```diff
@@ -1762,17 +1768,13 @@ TEST(Map, ObserveTileLifecycle) {
...
-        // Lower zooms can also be seen, but not always, so we
-        // ignore them.
-    };
-
     for (const auto& tile : expectedTiles) {
         TileOperation stage = TileOperation::NullOp;
+        bool parsing = false;
 
-        for (const auto& op : tileOps) {
+        // NOLINTNEXTLINE
```
Comment (louwers) [inline test/map/map.test.cpp:1775]: Maybe include the specific lint rule you are disabling?

**Review by louwers — APPROVED** [review]

File: src/mbgl/tile/geometry_tile.cpp (hunk 1 above)
Comment (mwilsnd) [reply src/mbgl/tile/geometry_tile.cpp]: Not really, but I don't think it ever truly was in the context of `reset()`. I've revised the comments in this method and cleared the `loaded` state which more accurately expresses what this method does.

File: test/map/map.test.cpp [hunk 3]
```diff
@@ -1772,6 +1772,7 @@ TEST(Map, ObserveTileLifecycle) {
         TileOperation stage = TileOperation::NullOp;
         bool parsing = false;
 
+        // Suppressing warning-as-error to use range-based for. Index is desired for debugging context.
         // NOLINTNEXTLINE
```
Comment (louwers) [inline test/map/map.test.cpp:1776]:
```suggestion
        // NOLINTNEXTLINE(modernize-loop-convert)
```

Like this I meant.

**Review by TimSylvester — APPROVED** [review]

**Review by alexcristici — APPROVED** [review]

**Pull request merged**